Feat/User query via RAG using LLM

ldw129 opened the pull request:
### Summary

Implement Retrieval-Augmented Generation (RAG) functionality for the user query feature in https://github.com/NotYuSheng/OmniPDF/issues/9

---

### Changes Made

<!-- List out major changes in this PR. Be brief but specific. -->

- Configure docker compose and env files for chat service to access the new persistent ChromaDB AsyncHTTPClient instance (https://github.com/NotYuSheng/OmniPDF/pull/114) running on port 8000 via API call
- Add new shared_utils module for the new persistent ChromaDB AsyncHTTPClient instance
- Configure Qwen2.5-0.5b-instruct as the primary LLM for initial RAG deployment, optimization and testing

---

### Context / Rationale

<!-- Why are these changes being made? Is it a bug fix, feature, refactor, etc.? Provide context to help reviewers understand your thought process. -->
- Configure docker compose and env files with the new persistent ChromaDB AsyncHTTPClient instance for chat service (https://github.com/NotYuSheng/OmniPDF/pull/114): To enable chat service to access the new persistent ChromaDB AsyncHTTPClient instance running on port 8000 via API call for the user query feature
- Add new shared_utils module for the new persistent ChromaDB AsyncHTTPClient instance: To enable the new persistent ChromaDB AsyncHTTPClient instance running on port 8000 to be accessible by embedder and chat services via API call
- Configure Qwen2.5-0.5b-instruct as the primary working LLM for initial RAG deployment, optimization and testing: To allow the user query feature to perform queries on the embeddings of data chunks in the new persistent ChromaDB AsyncHTTPClient instance and retrieve data accurately via RAG

---

### Related Docs or References

N.A.

---

### FastAPI Application Checklist (**Delete if PR is not relevant**)

- [x] API follows RESTful principles (nouns in routes, proper use of verbs)
- [x] All endpoints are async and use non-blocking I/O
- [x] `/health` endpoint is implemented and returns 200 OK
- [x] API schema (OpenAPI) is exposed and browsable at `/docs` or `/redoc`
- [x] Branch name follows conventions (e.g., `feature/*`, `bugfix/*`) — do **not** use `dev` directly

---

### General Checklist

- [x] I have tested these changes locally
- [x] I have updated relevant documentation or added comments where needed
- [x] I have linked relevant issues and tagged reviewers
- [x] I have followed coding conventions and naming standards

ldw129 committed:
fix: Update chat_service and chromadb in docker compose

ldw129 committed:
fix: Add openai, chromadb and sentence-transformers packages for chat service

ldw129 committed:
feat: Add new env variables for ChromaDB AsyncHTTPClient instance and Qwen-2.5 RAG model in .env file

ldw129 committed:
feat: Enhance ChatRequest and add ChatResponse models

ldw129 committed:
feat: Add config script with helper functions for Qwen-2.5 RAG model

ldw129 committed:
feat/perf: Add ChromaDB AsyncHTTPClient instance and RAG with Qwen-2.5 model functionality

ldw129 committed:
Revert "fix: Update chat_service and chromadb in docker compose"

This reverts commit ce8b595b93223641b53573887df53bbac8bf4217.

ldw129 committed:
fix: Update name of imported module "shared_utils.openai_client" in routers/chat

ldw129 committed:
perf: Import new chroma_client module from shared_utils folder to instantiate ChromaDB client instance in routers/chat

ldw129 committed:
perf: Add new chroma_client module in shared_utils folder

ldw129 committed:
chore: Remove unused import and commented out code from routers/chat

ldw129 committed:
chore: Adopt lowercase for boolean IS_PERSISTENT and ANONYMIZED_TELEMETRY env variables in chat's env file

ldw129 committed:
chore: Remove unnecessary constants and commented-out code from routers/chat

ldw129 committed:
chore: Remove unused serialize_chroma_results function and commented-out code from routers/chat

ldw129 committed:
chore: Remove unused import from routers/chat

ldw129 committed:
chore: Add new line at end of models/rag_config script based on PEP8 convention

ldw129 committed:
chore: Add new line at end of models/chat script based on PEP8 convention

ldw129 committed:
feat: Enhance RAG functionality to filter data based on which doc_id desired query data is located, and rank found doc_id's based on similarity_score

ldw129 committed:
Merge branch 'dev' into feat/chat_service

ldw129 committed:
fix: Update ChromaDB docker container in docker compose file

ldw129 committed:
chore: Add new line at end of shared_utils/chroma_client script based on PEP8 convention

ldw129 committed:
fix: Convert CHROMADB_PORT to int type for AsyncHTTPClient constructor

ldw129 committed:
fix: Update default_factory field for relevant_chunks parameter in ChatResponse model class

ldw129 committed:
fix: Update return path in perform_rag_query function at line 90 of routers/chat to accommodate 3 return values

ldw129 committed:
chore: Update return type hint for perform_rag_query function in routers/chat

ldw129 committed:
refactor: Remove unused parameter for post_process_qwen_response function in models/rag_config

ldw129 committed:
refactor: Remove unused parameter for rerank_chunks function in routers/chat

ldw129 commented:
/gemini review

ldw129 commented on shared_utils/chroma_client.py:
```diff
@@ -0,0 +1,13 @@
+    """
+    Initialize and return an AsyncHTTPClient ChromaDB client instance.
+    """
+    chroma_client = await chromadb.AsyncHttpClient(host=CHROMADB_HOST, port=int(CHROMADB_PORT))
```
My ChromaDB AsyncHTTPClient constructor is asynchronous in nature, not synchronous. Thus, I do need the await keyword.

ldw129 committed:
fix: Update default_factory field for relevant_chunks parameter in ChatResponse model class

ldw129 committed:
perf: Use get function with default value for parsing ChromaDB results in routers/chat for better robustness

ldw129 committed:
refactor: Remove unnecessary call to detect_query_type function for the query_type key of metadata dict in routers/chat

ldw129 committed:
chore: Remove unused ChromaDB env variables from env file of chat service

ldw129 committed:
chore: Remove unused chunk_overlap_threshold configuration property from rag_config

ldw129 commented:
/gemini review

ldw129 committed:
fix: Fix chunk reranking logic with Round-Robin approach for true interleaving

ldw129 committed:
refactor: Separate computation of number of documents from logging statement at line 102 in routers/chat

ldw129 committed:
refactor: Rename id parameter in ChatRequest model to doc_id

ldw129 commented:
/gemini review

ldw129 commented:
/gemini review

ldw129 committed:
fix: Remove similarity score sorting of reranked chunks

- Remove similarity score sorting of reranked chunks, refactor code for rerank_chunks function and update documentation in routers/chat
- Remove similarity score sorting of received reranked chunks for optimize_chunks_for_qwen function in models/rag_config

ldw129 committed:
perf: Add conditional checks for query type detection and response post processing in RAG by LLM

ldw129 commented:
/gemini review

ldw129 committed:
refactor: Define keyword lists for query type detection in QwenRAGOptimizer class as constant variables in model/rag_config

ldw129 committed:
refactor: Rename documents_searched field in metadata dict to documents_searched_count within handle_chat function in routers/chat

ldw129 commented:
/gemini review

ldw129 committed:
refactor: Update exception handling in routers/chat

ldw129 committed:
refactor: Remove unnecessary logging statements and exception handler in routers/chat

ldw129 committed:
fix: Remove return statement at line 92 in routers/chat

ldw129 commented:
/gemini review

ldw129 committed:
fix: Remove exception handling for rerank_chunks function in routers/chat

ldw129 committed:
fix: Update env variable used to call LLM in models/rag_config, and access model_name in QwenRAGConfig class from routers/chat

ldw129 committed:
refactor: Update type hints and return type for optimize_chunks_for_qwen function in models/rag_config

ldw129 committed:
fix: Re-define rerank_chunks as sync function and remove unused import in routers/chat

ldw129 committed:
Revert "fix: Re-define rerank_chunks as sync function and remove unused import in routers/chat"

This reverts commit 07332b327ad17a4cd060f3fa69158102bf6f705b.

ldw129 committed:
perf/chore: Add conditional check for relevant chunks in handle_chat function within routers/chat; Remove unused import

ldw129 committed:
fix: Change to AsyncOpenAI instance for openai_client in shared_utils

ldw129 commented:
/gemini review

ldw129 committed:
fix: Remove additional check for relevant_chunks

ldw129 committed:
refactor: Return a dictionary of prompt formats for format_user_prompt function and change to elif blocks for detect_query_type function in models/rag_config

ldw129 commented:
/gemini review

ldw129 committed:
fix: Fetch new AsyncOpenAI client instance from shared_utils

ldw129 committed:
perf: Add conditional check for relevant_chunks in handle_chat function within routers/chat

ldw129 committed:
fix: Remove HTTPException and return a ChatResponse model for empty relevant_chunks

ldw129 committed:
chore: Change to parentheses for keywords constant variables for immutability

ldw129 commented:
/gemini review

ldw129 committed:
refactor: Replace 'unknown' default key value with None

ldw129 commented:
@NotYuSheng ready to merge, code base locally tested successfully together with PR #114, with no issues connecting to LMStudio and ChromaDB

ldw129 committed:
Merge branch 'dev' into feat/chat_service

NotYuSheng commented on chat_service/models/rag_config.py:
```diff
@@ -0,0 +1,183 @@
+        
+        return prompt_formats.get(query_type, prompt_formats["general"])
+
+class QwenRAGOptimizer:
+    """Advanced optimization techniques for Qwen-2.5 RAG"""
+    
+    @staticmethod
+    def detect_query_type(question: str) -> str:
+        """Automatically detect the type of query to use appropriate prompting"""
+        
+        question_lower = question.lower()
+        
+        # Factual indicators
+        if any(keyword in question_lower for keyword in FACTUAL_KEYWORDS):
+            return "factual"
+        
+        # Analytical indicators
+        elif any(keyword in question_lower for keyword in ANALYTICAL_KEYWORDS):
+            return "analytical"
+        
+        # Summarization indicators
+        elif any(keyword in question_lower for keyword in SUMMARY_KEYWORDS):
+            return "summarization"
+        
+        return "general"
```
Problematic hard coding, consider switching to an agentic approach

NotYuSheng reviewed (changes requested):
Address comment please

ldw129 committed:
feat: Enhance RAG functionality, add query validation for LLM to decide on performing RAG and do away query type classification based on keywords

ldw129 commented:
/gemini review

ldw129 committed:
refactor: Add parenthesis to elif condition in line 210

ldw129 committed:
refactor: Initialize collection_info to None before conditional block at line 178

ldw129 committed:
refactor: Add fallback return for detect_query_type method in QwenRAGOptimizer class in rag_config

ldw129 committed:
chore: Remove unused commented-out code in routers/chat

ldw129 committed:
refactor: Update 3f8f51e and 4f2c51c to resolve code errors

ldw129 commented:
@gemini-code-assist can you help me to refine my logic for validate_query_with_llm such that my LLM used for the RAG can analyze the nature of the query more accurately so as to make a more informed decision on whether to perform RAG from chromaDB?

ldw129 committed:
feat: Enhance context of ChromaDB collection as part of system prompt for LLM to do RAG

ldw129 committed:
refactor: Remove unnecessary else block at line 181 in routers/chat

ldw129 committed:
refactor: Replace generic exception handling to APIError for LLM interaction

ldw129 commented:
/gemini review

ldw129 committed:
fix: Rename env variable under RAG Optimization in example.env

ldw129 committed:
fix: Include validation_error in failed ChatResponse response

ldw129 committed:
fix: Remove None initialization for openai_client in rag_config

ldw129 commented:
/gemini review

ldw129 committed:
perf: Enhance robustness of LLM decision-making on RAG

ldw129 commented:
/gemini review

ldw129 commented on chat_service/routers/chat.py:
```diff
@@ -1,58 +1,365 @@
+        
+        if not should_rag:
+            logger.info(f"LLM query validation failed: {validation_error}")
+            return ChatResponse(
+                response=f"""I'm sorry, but I couldn't process your query based on the documents in {chat_request.collection_name} that you want to query data from. {validation_error} Please provide a clear, specific question that I can help answer using the available documents in {chat_request.collection_name}.""",
+                relevant_chunks=[],
+                metadata={
+                    "query_type": "invalid",
+                    "chunks_used": 0,
+                    "documents_searched_count": 0,
+                    "document_ids": [],
+                    "total_context_length": 0,
+                    "model_used": OPENAI_MODEL_NAME,
+                    "collection_name": chat_request.collection_name,
+                    "validation_error": validation_error,
+                    "validation_method": "llm",
+                    "rag_performed": False,
+                    "generation_params": qwen_config.generation_params,
+                    "reranking_enabled": qwen_config.enable_reranking
+                }
+            )
```
I will standardize the fields to be provided in the metadata.

ldw129 committed:
refactor: Standardize metadata fields in ChatResponse model for better code maintainability

ldw129 committed:
chore: Remove unused commented-out code

ldw129 committed:
chore: Update documentation in routers/chat

ldw129 committed:
fix: Fix metadata dictionary for ChatResponse model

ldw129 commented:
/gemini review

ldw129 committed:
refactor: Remove top_k parameter from ChatRequest model and declare top_k as new env variable in env file for chat service

ldw129 committed:
refactor: Simplify for loop for processing ChromaDB results in prepare_retrieval_results function in routers/chat

ldw129 commented:
/gemini review

ldw129 committed:
fix: Initialize top_k param as int instead of float in rag_config

ldw129 commented on chat_service/routers/chat.py:
```diff
@@ -1,58 +1,344 @@
+            **qwen_config.validation_params
+        )
+        
+        result = response.choices[0].message.content.strip()
+        logger.info(f"Result: {result}")
+        decision_line = next((line for line in result.split('\n') if line.startswith("DECISION:")), "")
+        
+        if "PROCEED_WITH_RAG" in decision_line:
+            return True, None
+
+        return False, result
```
@gemini-code-assist how can I do so?

ldw129 committed:
fix: Use JSON for LLM's validation decision

- Enhance LLM's response with JSON
- Move json import to next line in routers/chat
- Use JSON method to check LLM's validation decision

ldw129 commented:
/gemini review

ldw129 committed:
chore: Remove unused QueryValidationResult class from rag_config

ldw129 committed:
fix: Return json_result if PROCEED_WITH_RAG_ and update response param in ChatResponse to be returned in routers/chat

ldw129 commented on chat_service/routers/chat.py:
```diff
@@ -1,58 +1,352 @@
+        chunk={
+            'content': doc,
+            'chunk_id': chunk_id,
+            'similarity_score': 1 - distance, # Convert distance to similarity score
```
@gemini-code-assist ChromaDB also has cosine as one of the acceptable distance metrics, apart from l2 and Inner Product (ip).

ldw129 replied:
@gemini-code-assist how can I modify my code based on your explanation?

ldw129 committed:
fix: Change formula for similarity_score in routers/chat

ldw129 committed:
Remove default None values for function params

- Remove default None value for model_name and config params in detect_query_type function in rag_config
- Remove default None value for openai_client param in perform_rag_query and _validate_query_with_llm functions in routers/chat

ldw129 commented:
/gemini review

ldw129 committed:
fix: Change type-hint to return tuple[bool, Dict[str, Any]] for validate_query_with_llm function in routers/chat

ldw129 committed:
fix: Remove leading underscore from function name

ldw129 commented:
/gemini review

ldw129 committed:
fix: Update JSONDecodeError handling in routers/chat

ldw129 commented:
/gemini review

ldw129 commented:
@NotYuSheng ready to be merged

ldw129 committed:
chore: Add LLM_API_URL to openai_client

ldw129 committed:
fix: Revert JSON formatting for LLM's decision-making on RAG

ldw129 commented:
/gemini review

ldw129 committed:
fix: Fix typo in validation prompt

ldw129 commented:
/gemini review

ldw129 committed:
Revert "fix: Revert JSON formatting for LLM's decision-making on RAG"

This reverts commit 02fe104831f9a92cb16937fe95c941ef627240de.

ldw129 committed:
Revert "fix: Change type-hint to return tuple[bool, Dict[str, Any]] for validate_query_with_llm function in routers/chat"

This reverts commit 355b7af13c39842e480258aeff7b40f592c7a37c.

ldw129 committed:
Revert "Remove default None values for function params"

This reverts commit c480925e00065fcae2e49c1bad78e52f04fca1cf.

ldw129 committed:
Revert "fix: Return json_result if PROCEED_WITH_RAG_ and update response param in ChatResponse to be returned in routers/chat"

This reverts commit 6b3c627e7c9aacc405081472d00d2b00421025de.

ldw129 committed:
Revert "fix: Use JSON for LLM's validation decision"

This reverts commit 49d3d8a127cdc5f368ae7c2c652e088423a9a543.

ldw129 committed:
Revert "fix: Remove None initialization for openai_client in rag_config"

This reverts commit 51f49fe88c8c654cf27e67b83a1bad00dcd9d684.

ldw129 committed:
fix: Add a PROCEED_WITH_RAG example and replace top_k param with top_p param in generation_params for rag_config

ldw129 committed:
fix: Remove : in LLM_API_URL in example.env

ldw129 committed:
refactor: Initialize QWEN_TOP_K variable

ldw129 commented:
/gemini review

ldw129 committed:
perf: Enhance post_process_qwen_response logic

ldw129 committed:
refactor: Remove redundant check on llm_classification

ldw129 committed:
chore: Remove unused call to OPENAI_BASE_URL

MingJ7 commented on chat_service/example.env:
```diff
@@ -1,3 +1,21 @@
 OPENAI_API_KEY=lm-studio
+
+# Qwen-2.5 Model Configuration
 OPENAI_MODEL=qwen2.5-0.5b-instruct
```
Nit: Please change the model to qwen2.5 14b

MingJ7 commented on chat_service/models/chat.py:
```diff
@@ -8,4 +8,14 @@ class ChatRequest(BaseModel):
     message: str
-    id: Optional[str] = None
+    doc_id: Optional[str] = None
+    collection_name: str = Field(default="default_collection", description="ChromaDB collection name")
```
Suggest: Consider removing this field as we plan to move into the 2 collection structure.

ldw129 replied:
To be done in a future PR

ldw129 committed:
Add a tag to /chat endpoint for consistency

ldw129 committed:
fix: Update OPENAI_MODEL in example.env

NotYuSheng commented on chat_service/example.env:
```diff
@@ -1,3 +1,21 @@
 OPENAI_BASE_URL=http://localhost:1234/v1 # Please change this to your LM Studio URL
+LLM_API_URL=http://webworkdgx/vllm_qwen3coder/v1 # DGX URL
```
Why is there two API endpoints for llm server? this should be the same thing

NotYuSheng replied:
I suggest removing LLM_API_URL and replacing it with OPENAI_BASE_URL

I recommend using this:
```
OPENAI_BASE_URL=http://webworkdgx/vllm_qwen3coder/v1 # Please change this to your LLM Server URL endpoint
```

ldw129 replied:
[fixed ](https://github.com/NotYuSheng/OmniPDF/pull/115/commits/4a2238e47d7f1efecf1e07f4052723186344f216)

NotYuSheng replied:
OPENAI_BASE_URL=http://webworkdgx/vllm_qwen3coder/v1 # Please change this to your LLM Server URL endpoint


please keep the comment

NotYuSheng commented on chat_service/models/rag_config.py:
Refactor required

Note that qwen will likely not be the actual model we use in production. Therefore the program should not directly reference qwen.

NotYuSheng replied:
Since this is called by the chat.py, be sure to refactor there as well.

NotYuSheng replied:
```
            QueryType.FACTUAL.value: """You are Qwen, a precision-focused AI assistant for factual document analysis.
```

For these lines, I believe something like this should work:
```
            QueryType.FACTUAL.value: """You are a precision-focused AI assistant for factual document analysis.
```

NotYuSheng commented on shared_utils/openai_client.py:
```diff
@@ -1,12 +1,12 @@
-    return OpenAI(
-        base_url=os.getenv("OPENAI_BASE_URL"),  # Make sure `/v1` is included
+    return AsyncOpenAI(
+        base_url=os.getenv("LLM_API_URL"), # Make sure `/v1` is included
```
Switch back

NotYuSheng commented on chat_service/routers/chat.py:
Refactor to remove references to qwen

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/115/commits/41612d8d2901ceb523d8d31020995a4e5549852c)

NotYuSheng reviewed (changes requested):
Please adjust based on feedback

ldw129 committed:
fix: Remove additional LLM URL used for local testing from env and openai_client

ldw129 committed:
refactor: Remove all direct references to Qwen for code scalability

NotYuSheng commented on chat_service/models/rag_config.py:
```diff
@@ -0,0 +1,505 @@
+    """Configuration class for the model's RAG optimization"""
+    
+    def __init__(self):
+        self.model_name = os.getenv("OPENAI_MODEL", "Qwen2.5-14B-Coder-Instruct")
```
Please remove the default reference

ldw129 replied:
[fixed](https://github.com/NotYuSheng/OmniPDF/pull/115/commits/055b99d90078db5dd8a4d0ce338e4778fe93fb81)

ldw129 committed:
fix: Update comment for OPENAI_BASE_URL in env file

ldw129 committed:
refactor: Remove default reference for OPENAI_MODEL in rag_config

ldw129 committed:
fix: Update comment for OPENAI_BASE_URL in env file

NotYuSheng reviewed (changes requested):
last 2 change before merge

NotYuSheng reviewed (approved):
Approved and merge!
